internal/container/container.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package container
+
+import (
+	"context"
+	"io"
+	"syscall"
+
+	"github.com/containerd/containerd/cio"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/fifo"
+	"github.com/gogo/protobuf/types"
+	ptypes "github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal/bundle"
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+)
+
+var (
+	_ VMTask = &vmTask{}
+)
+
+type VMTask interface {
+	Pack() (*ptypes.Any, error)
+	RuncOptions() *ptypes.Any
+	ProxyStdio(StdioConnector) error
+	Cancel()
+}
+
+type vmTask struct {
+	id          string
+	bundleDir   bundle.Dir
+	runcOptions *ptypes.Any
+	fifoSet     *cio.FIFOSet
+
+	stdinPort  uint32
+	stdoutPort uint32
+	stderrPort uint32
+
+	ctx    context.Context
+	cancel func()
+}
+
+func New(
+	containerID string,
+	bundleDir bundle.Dir,
+	options *ptypes.Any,
+	fifoSet *cio.FIFOSet,
+	stdinPort, stdoutPort, stderrPort uint32,
+) VMTask {
+	ctrCtx, cancel := context.WithCancel(context.Background())
+
+	return &vmTask{
+		id:          containerID,
+		bundleDir:   bundleDir,
+		runcOptions: options,
+		fifoSet:     fifoSet,
+
+		stdinPort:  stdinPort,
+		stdoutPort: stdoutPort,
+		stderrPort: stderrPort,
+
+		ctx:    ctrCtx,
```

Comment:
nit: having `ctx` as a struct field is not not recommended [practice](https://dave.cheney.net/2017/08/20/context-isnt-for-cancellation). Is there any reason why  we can't pass it as func parameter instead?

Reply:
That makes sense. I didn't want to necessarily use the context that is passed in everywhere because the container exists longer than the lifetime of any single request (i.e. i/o copying lasts for the whole lifetime of the container).

I will refactor this to use the passed in context where it makes sense and use a non-context-based way of canceling the I/O (or something along those lines)